### PR TITLE
[WIP] Start to sketch out integration with `cosign`.

### PR DESCRIPTION
**Pull request by mattmoor**
When `COSIGN_EXPERIMENTAL=true`, this will verify base images, and sign produced images using the KEYLESS flow.

Fixes: https://github.com/google/ko/issues/357
Fixes: https://github.com/google/ko/issues/356

### CODE REVIEW

**Comment by mattmoor** [comment]
`cosign` uses Go 1.16 features, so we should discuss what to do with some of the older tests 🤔

File: pkg/commands/config.go [hunk 1]
```diff
@@ -124,6 +127,18 @@ func getBaseImage(platform string, bo *options.BuildOptions) build.GetBase {
 		if err != nil {
 			return nil, nil, err
 		}
+
+		// With the resolved digest in hand (in descriptor), we verify the
+		// signature of the base image.
+		if cli.EnableExperimental() {
+			digest := ref.Context().String() + "@" + desc.Digest.String()
```
Comment (imjasonh) [inline pkg/commands/config.go:134]: `ref.Context().Digest(desc.Digest.String())`

File: pkg/commands/config.go [hunk 2]
```diff
@@ -139,6 +154,52 @@ func getBaseImage(platform string, bo *options.BuildOptions) build.GetBase {
...
+}
+
+// entry is used to hold the sync.Once used to execute the digest verification,
+// and the resulting error (possibly nil), that results from verification.
+type entry struct {
+	once   sync.Once
+	result error
+}
+
+func (v *verifier) verify(ctx context.Context, digest string) error {
```
Comment (imjasonh) [inline pkg/commands/config.go:171]: Can `digest` be a `name.Digest` instead of a string?

File: pkg/commands/config.go [hunk 3]
```diff
@@ -139,6 +154,52 @@ func getBaseImage(platform string, bo *options.BuildOptions) build.GetBase {
...
+	e := v.getEntry(digest)
+	e.once.Do(func() {
+		// cosign emits a lot of stuff to STDOUT that we don't want
```
Comment (imjasonh) [inline pkg/commands/config.go:174]: I know this is just WIP but this is gross, we should fix this upstream before papering over it here.

**Review by imjasonh — COMMENTED** [review]
I'm honestly pleaseantly surprised how relatively small this diff is, in terms of ko lines changed. However, we should try to refactor cosign's packages to make them easier for us (and others, hopefully) to consume in this way, so this diff isn't so omghuge.

File: pkg/commands/config.go (hunk 1 above)
Comment (mattmoor) [reply pkg/commands/config.go:134]: neat, TIL

File: pkg/commands/config.go (hunk 2 above)
Comment (mattmoor) [reply pkg/commands/config.go:171]: yup

File: pkg/commands/config.go (hunk 3 above)
Comment (mattmoor) [reply pkg/commands/config.go:174]: I did, it broke and got rolled back.  I need to sort out what needs to be fixed in my fix :)

**Comment by mattmoor** [comment]
> I'm honestly pleaseantly surprised how relatively small this diff is

Yeah, I've been working to minimize it 😉 

Really aside from the output garbage and the logic to eliminate redundant verifications it's like 5 LoC each.

> we should try to refactor cosign's packages to make them easier for us (and others, hopefully) to consume in this way

Part of me wonders if we'd have even noticed if we didn't check in `vendor`.  I agree, and would certainly appreciate any help in that department.  🙏 😇

**Review by jonjohnsonjr — CHANGES_REQUESTED** [review]
ಠ_ಠ

**Comment by mattmoor** [comment]
@jonjohnsonjr @imjasonh https://github.com/sigstore/cosign/issues/657 👀

**Comment by mattmoor** [comment]
@dlorenc thinks I should play with the stuff in `pkg/cosign` instead.  I'll see how much leaner that is from a `vendor` perspective, and how much more code it requires.

File: pkg/commands/config.go (hunk 3 above)
Comment (mattmoor) [reply pkg/commands/config.go]: This is gone from the latest update.  Inlining stuff FTW, I guess.

**Comment by imjasonh** [comment]
Needs rebase.

**Pull request closed without merging**